File: engines/android-app/app/src/main/kotlin/io/saturnis/trajectory/manager/WorkflowManager.kt
// Copyright (c) 2026 Dennis Brandl
// Licensed under the Apache License, Version 2.0. See LICENSE for details.
package io.saturnis.trajectory.manager

import io.saturnis.trajectory.coordinator.WorkflowCoordinator
import io.saturnis.trajectory.coordinator.CoordinatorState
import io.saturnis.trajectory.storage.WorkflowDao
import io.saturnis.trajectory.storage.LoadedWorkflowEntity
import io.saturnis.trajectory.storage.ActiveWorkflowEntity
import io.saturnis.trajectory.storage.CompletedWorkflowEntity
import com.trajectoryruntime.engine.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.encodeToString
import java.io.File
import java.util.UUID

data class LoadedWorkflow(
    val id: String,
    val specOid: String,
    val localId: String,
    val version: String,
    val name: String,
    val spec: MasterWorkflowSpecification,
    val mediaMap: Map<String, String>,
    val mediaDir: String?,
    val environmentJsons: List<String> = emptyList(),
    val loadedAt: Long,
)

data class ActiveWorkflow(
    val id: String,
    val name: String,
    val localId: String,
    val version: String,
    val coordinator: WorkflowCoordinator,
    val specJson: String,
    val mediaMapJson: String?,
    val environmentsJson: String?,
    val startedAt: Long,
)

data class CompletedWorkflow(
    val id: String,
    val localId: String,
    val version: String,
    val name: String,
    val workflowState: String,
    val startedAt: Long,
    val finishedAt: Long,
    val traceJson: String,
    val propertiesJson: String,
    val specJson: String,
)

data class ManagerState(
    val loaded: List<LoadedWorkflow> = emptyList(),
    val active: List<ActiveWorkflow> = emptyList(),
    val completed: List<CompletedWorkflow> = emptyList(),
    val focusedActiveId: String? = null,
)

sealed class AddResult {
    data class Success(val id: String) : AddResult()
    data class Duplicate(val existingId: String) : AddResult()
}

class WorkflowManager(
    private val dao: WorkflowDao?,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main),
) {
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }
    private val _state = MutableStateFlow(ManagerState())
    val state: StateFlow<ManagerState> = _state.asStateFlow()

    private val coordinatorJobs = mutableMapOf<String, Job>()
    /** Shared resource managers for environment-scoped cross-workflow sync, keyed by env OID set. */
    private val envResourceManagers = mutableMapOf<String, InMemoryResourceManager>()
    /** Guard against re-entrant pumping of sibling coordinators. */
    private var pumpingInProgress = false

    fun addWorkflow(
        spec: MasterWorkflowSpecification,
        mediaMap: Map<String, String>,
        mediaDir: String? = null,
        environmentJsons: List<String> = emptyList(),
    ): AddResult {
        val existing = _state.value.loaded.find { it.specOid == spec.oid }
        if (existing != null) return AddResult.Duplicate(existing.id)

        val id = UUID.randomUUID().toString()
        val loaded = LoadedWorkflow(
            id = id, specOid = spec.oid, localId = spec.local_id,
            version = spec.version, name = spec.local_id,
            spec = spec, mediaMap = mediaMap, mediaDir = mediaDir,
            environmentJsons = environmentJsons,
            loadedAt = System.currentTimeMillis(),
        )

        _state.update { it.copy(loaded = it.loaded + loaded) }

        dao?.let { d ->
            scope.launch(Dispatchers.IO) {
                d.insertLoaded(LoadedWorkflowEntity(
                    id = id, localId = spec.local_id, version = spec.version,
                    name = loaded.name, specOid = spec.oid,
                    specJson = json.encodeToString(spec),
                    mediaDir = mediaDir,
                    mediaMapJson = json.encodeToString(mediaMap),
                    loadedAt = loaded.loadedAt,
                ))
            }
        }

        return AddResult.Success(id)
    }

    @OptIn(kotlinx.coroutines.FlowPreview::class)
    fun startWorkflow(loadedId: String, startingParams: Map<String, String>? = null, allowScriptExecution: Boolean = false): String? {
        val loaded = _state.value.loaded.find { it.id == loadedId } ?: return null
        val instanceId = UUID.randomUUID().toString()
        val coordinator = WorkflowCoordinator()

        val specJsonStr = json.encodeToString(loaded.spec)
        val mediaMapJsonStr = json.encodeToString(loaded.mediaMap)
        val envsJsonStr = if (loaded.environmentJsons.isNotEmpty()) json.encodeToString(loaded.environmentJsons) else null

        val active = ActiveWorkflow(
            id = instanceId,
            name = loaded.name, localId = loaded.localId,
            version = loaded.version, coordinator = coordinator,
            specJson = specJsonStr, mediaMapJson = mediaMapJsonStr,
            environmentsJson = envsJsonStr,
            startedAt = System.currentTimeMillis(),
        )

        _state.update { it.copy(
            active = it.active + active,
            focusedActiveId = instanceId,
        ) }

        // Persist active to Room BEFORE starting
        dao?.let { d ->
            scope.launch(Dispatchers.IO) {
                d.insertActive(ActiveWorkflowEntity(
                    id = instanceId,
                    name = loaded.name, localId = loaded.localId,
                    version = loaded.version,
                    specJson = specJsonStr,
                    mediaMapJson = mediaMapJsonStr,
                    environmentsJson = envsJsonStr,
                    traceJson = "[]", propertiesJson = "{}",
                    userActionsJson = "[]", workflowState = "RUNNING",
                    activeStepsJson = "[]",
                    startedAt = active.startedAt,
                    lastUpdatedAt = active.startedAt,
                ))
            }
        }

        // Start engine (synchronous — state settles immediately)
        val setup = if (startingParams != null) TestFixtureSetup(starting_parameters = startingParams) else null
        val sharedRM = getSharedResourceManager(loaded.environmentJsons)
        coordinator.loadAndStart(loaded.spec, setup = setup, mediaMap = loaded.mediaMap, environmentJsons = loaded.environmentJsons, sharedResourceManager = sharedRM, allowScriptExecution = allowScriptExecution)

        // After start, pump siblings in case this workflow's activation granted resources for others
        pumpSiblingCoordinators(instanceId)

        // Check if already completed (e.g., START→END workflow)
        val currentState = coordinator.state.value.workflowState
        if (currentState in setOf(
            WorkflowState.COMPLETED, WorkflowState.ABORTED,
            WorkflowState.STOPPED, WorkflowState.ERRORED,
        )) {
            completeWorkflow(instanceId, coordinator.state.value)
            return instanceId
        }

        // Watch for future completion and pump cross-workflow resource grants
        val job = scope.launch {
            coordinator.state.collect { coordState ->
                pumpSiblingCoordinators(instanceId)
                if (coordState.workflowState in setOf(
                    WorkflowState.COMPLETED, WorkflowState.ABORTED,
                    WorkflowState.STOPPED, WorkflowState.ERRORED,
                )) {
                    completeWorkflow(instanceId, coordState)
                    cancel()
                }
            }
        }
        coordinatorJobs[instanceId] = job

        // Persist snapshots on state changes (debounced 300ms)
        val snapshotJob = scope.launch {
            coordinator.state
                .drop(1) // skip current value
                .debounce(300)
                .conflate()
                .collect { coordState ->
                    persistActiveSnapshot(instanceId, coordinator, coordState)
                }
        }
        coordinatorJobs["${instanceId}-snapshot"] = snapshotJob

        return instanceId
    }

    /** Get or create a shared resource manager for workflows sharing environments. */
    private fun getSharedResourceManager(environmentJsons: List<String>): InMemoryResourceManager? {
        if (environmentJsons.isEmpty()) return null
        // Build stable key from sorted environment OIDs
        val envOids = mutableListOf<String>()
        for (envJson in environmentJsons) {
            try {
                val lib = json.parseToJsonElement(envJson).jsonObject
                val specs = lib["environment_specifications"]?.jsonArray ?: continue
                for (specEl in specs) {
                    val oid = specEl.jsonObject["oid"]?.jsonPrimitive?.contentOrNull ?: continue
                    envOids.add(oid)
                }
            } catch (e: Exception) {
                android.util.Log.w("WorkflowManager", "Failed to parse environment JSON", e)
            }
        }
        if (envOids.isEmpty()) return null
        val key = envOids.sorted().joinToString(",")
        return envResourceManagers.getOrPut(key) { InMemoryResourceManager() }
    }

    /** Pump all active coordinators (except the source) to check for cross-workflow resource grants. */
    private fun pumpSiblingCoordinators(sourceId: String) {
        if (pumpingInProgress) return
        pumpingInProgress = true
        try {
            for (wf in _state.value.active) {
                if (wf.id == sourceId) continue
                wf.coordinator.pumpSharedResources()
            }
        } finally {
            pumpingInProgress = false
        }
    }

    fun removeLoadedWorkflow(id: String) {
        val loaded = _state.value.loaded.find { it.id == id }
        _state.update { it.copy(loaded = it.loaded.filter { w -> w.id != id }) }
        dao?.let { d ->
            scope.launch(Dispatchers.IO) {
                d.deleteLoaded(id)
                // Clean up extracted files from disk
                loaded?.mediaDir?.let { dir ->
                    File(dir).deleteRecursively()
                }
            }
        }
    }

    fun removeCompletedWorkflow(id: String) {
        _state.update { it.copy(completed = it.completed.filter { w -> w.id != id }) }
        dao?.let { d -> scope.launch(Dispatchers.IO) { d.deleteCompleted(id) } }
    }

    fun focusWorkflow(id: String) {
        _state.update { it.copy(focusedActiveId = id) }
    }

    fun releaseAllEnvironmentResources(): List<String> {
        val released = mutableSetOf<String>()
        for (active in _state.value.active) {
            released.addAll(active.coordinator.releaseAllEnvironmentResources())
        }
        return released.toList()
    }

    fun releaseAllWorkflowResources(): List<String> {
        val released = mutableSetOf<String>()
        for (active in _state.value.active) {
            released.addAll(active.coordinator.releaseAllWorkflowResources())
        }
        return released.toList()
    }

    /** Aggregate env-resource snapshot across all active workflows, deduplicated by
     *  composite resourceKey (`name` field). When the same env-scoped pool is shared
     *  by multiple workflows, the entry reflects current pool state once. */
    fun getEnvironmentResourceSnapshot(): List<com.trajectoryruntime.engine.ResourceSnapshotEntry> {
        val seen = linkedMapOf<String, com.trajectoryruntime.engine.ResourceSnapshotEntry>()
        for (active in _state.value.active) {
            for (e in active.coordinator.getEnvironmentResourceSnapshot()) {
                seen.putIfAbsent(e.name, e)
            }
        }
        return seen.values.toList()
    }

    /** Aggregate workflow-scoped resource snapshot across all active workflows. */
    fun getWorkflowResourceSnapshot(): List<com.trajectoryruntime.engine.ResourceSnapshotEntry> {
        val seen = linkedMapOf<String, com.trajectoryruntime.engine.ResourceSnapshotEntry>()
        for (active in _state.value.active) {
            for (e in active.coordinator.getWorkflowResourceSnapshot()) {
                seen.putIfAbsent(e.name, e)
            }
        }
        return seen.values.toList()
    }

    /** Reset a single resource by composite resourceKey. Tries each coordinator
     *  until one reports the resource exists in its manager view. */
    fun resetResource(resourceKey: String): Boolean {
        for (active in _state.value.active) {
            if (active.coordinator.resetResource(resourceKey)) return true
        }
        return false
    }

    fun getCoordinator(activeId: String): WorkflowCoordinator? {
        return _state.value.active.find { it.id == activeId }?.coordinator
    }

    private fun completeWorkflow(instanceId: String, finalState: CoordinatorState) {
        android.util.Log.d("WorkflowManager", "completeWorkflow called: id=$instanceId state=${finalState.workflowState}")
        val active = _state.value.active.find { it.id == instanceId }
        if (active == null) {
            android.util.Log.w("WorkflowManager", "completeWorkflow: workflow $instanceId not found in active list (already completed?)")
            return
        }

        coordinatorJobs.remove(instanceId)?.cancel()
        coordinatorJobs.remove("${instanceId}-snapshot")?.cancel()

        val completed = CompletedWorkflow(
            id = instanceId, localId = active.localId,
            version = active.version, name = active.name,
            workflowState = finalState.workflowState.name,
            startedAt = active.startedAt,
            finishedAt = System.currentTimeMillis(),
            traceJson = json.encodeToString(finalState.trace),
            propertiesJson = json.encodeToString(finalState.properties),
            specJson = json.encodeToString(active.coordinator.getSpec()),
        )

        _state.update { current ->
            val newActive = current.active.filter { it.id != instanceId }
            android.util.Log.d("WorkflowManager", "completeWorkflow: active ${current.active.size} -> ${newActive.size}, completed ${current.completed.size} -> ${current.completed.size + 1}")
            current.copy(
                active = newActive,
                completed = listOf(completed) + current.completed,
                focusedActiveId = if (current.focusedActiveId == instanceId) {
                    newActive.firstOrNull()?.id
                } else current.focusedActiveId,
            )
        }

        active.coordinator.reset()

        dao?.let { d ->
            scope.launch(Dispatchers.IO) {
                d.deleteActive(instanceId)
                d.insertCompleted(CompletedWorkflowEntity(
                    id = completed.id, localId = completed.localId,
                    version = completed.version, name = completed.name,
                    specJson = completed.specJson,
                    traceJson = completed.traceJson,
                    propertiesJson = completed.propertiesJson,
                    workflowState = completed.workflowState,
                    startedAt = completed.startedAt,
                    finishedAt = completed.finishedAt,
                ))
            }
        }
    }

    private suspend fun persistActiveSnapshot(
        instanceId: String,
        coordinator: WorkflowCoordinator,
        coordState: CoordinatorState,
    ) {
        dao?.let { d ->
            withContext(Dispatchers.IO) {
                val active = _state.value.active.find { it.id == instanceId } ?: return@withContext
                d.updateActive(ActiveWorkflowEntity(
                    id = instanceId,
                    name = active.name, localId = active.localId,
                    version = active.version,
                    specJson = active.specJson,
                    mediaMapJson = active.mediaMapJson,
                    environmentsJson = active.environmentsJson,
                    traceJson = json.encodeToString(coordState.trace),
                    propertiesJson = json.encodeToString(coordState.properties),
                    userActionsJson = json.encodeToString(coordinator.getUserActions()),
                    workflowState = coordState.workflowState.name,
                    activeStepsJson = "[]",
                    startedAt = active.startedAt,
                    lastUpdatedAt = System.currentTimeMillis(),
                ))
            }
        }
    }

    suspend fun resumeFromDatabase() {
        val d = dao ?: return
        withContext(Dispatchers.IO) {
            val loadedEntities = d.observeLoaded().first()
            val loadedList = loadedEntities.mapNotNull { entity ->
                try {
                    val spec = json.decodeFromString<MasterWorkflowSpecification>(entity.specJson)
                    val mediaMap: Map<String, String> = entity.mediaMapJson?.let {
                        json.decodeFromString(it)
                    } ?: emptyMap()
                    LoadedWorkflow(
                        id = entity.id, specOid = entity.specOid,
                        localId = entity.localId, version = entity.version,
                        name = entity.name, spec = spec, mediaMap = mediaMap,
                        mediaDir = entity.mediaDir, loadedAt = entity.loadedAt,
                    )
                } catch (e: Exception) {
                    android.util.Log.e("WorkflowManager", "Failed to restore loaded workflow ${entity.id}", e)
                    null
                }
            }

            val activeEntities = d.getAllActive()
            val activeList = activeEntities.mapNotNull { entity ->
                try {
                    val spec = json.decodeFromString<MasterWorkflowSpecification>(entity.specJson)
                    val mediaMap: Map<String, String> = entity.mediaMapJson?.let {
                        json.decodeFromString(it)
                    } ?: emptyMap()
                    val envJsons: List<String> = entity.environmentsJson?.let {
                        json.decodeFromString(it)
                    } ?: emptyList()
                    val coordinator = WorkflowCoordinator()
                    val userActions: List<UserAction> = json.decodeFromString(entity.userActionsJson)
                    val sharedRM = getSharedResourceManager(envJsons)
                    coordinator.loadAndStart(spec, mediaMap = mediaMap, environmentJsons = envJsons, sharedResourceManager = sharedRM)
                    for (action in userActions) {
                        coordinator.submitAction(action)
                    }
                    ActiveWorkflow(
                        id = entity.id,
                        name = entity.name, localId = entity.localId,
                        version = entity.version, coordinator = coordinator,
                        specJson = entity.specJson,
                        mediaMapJson = entity.mediaMapJson,
                        environmentsJson = entity.environmentsJson,
                        startedAt = entity.startedAt,
                    )
                } catch (e: Exception) {
                    android.util.Log.e("WorkflowManager", "Failed to restore active workflow ${entity.id}", e)
                    d.deleteActive(entity.id)
                    null
                }
            }

            val completedEntities = d.observeCompleted().first()
            val completedList = completedEntities.map { entity ->
                CompletedWorkflow(
                    id = entity.id, localId = entity.localId,
                    version = entity.version, name = entity.name,
                    workflowState = entity.workflowState,
                    startedAt = entity.startedAt, finishedAt = entity.finishedAt,
                    traceJson = entity.traceJson,
                    propertiesJson = entity.propertiesJson,
                    specJson = entity.specJson,
                )
            }

            withContext(Dispatchers.Main) {
                _state.value = ManagerState(
                    loaded = loadedList,
                    active = activeList,
                    completed = completedList,
                    focusedActiveId = activeList.firstOrNull()?.id,
                )
            }
        }
    }
}

File: engines/web-ui/src/manager/WorkflowManager.ts
// Copyright (c) 2026 Dennis Brandl
// Licensed under the Apache License, Version 2.0. See LICENSE for details.
import type {
  MasterWorkflowSpecification,
  MasterEnvironmentLibrary,
} from '@engine/types.js';
import { deepCopySpec } from '@engine/spec-copy.js';
import { InMemoryResourceManager } from '@engine/resource-manager.js';
import { WorkflowCoordinator } from '../coordinator/WorkflowCoordinator';
import type { LoadedWorkflow, ActiveWorkflow, CompletedWorkflow, ManagerSnapshot, Listener } from './types';

/** crypto.randomUUID() requires a secure context (HTTPS/localhost).
 *  This fallback uses crypto.getRandomValues() which works over plain HTTP. */
function generateUUID(): string {
  if (typeof crypto !== 'undefined' && crypto.randomUUID) {
    return crypto.randomUUID();
  }
  const bytes = new Uint8Array(16);
  crypto.getRandomValues(bytes);
  bytes[6] = (bytes[6] & 0x0f) | 0x40; // version 4
  bytes[8] = (bytes[8] & 0x3f) | 0x80; // variant 1
  const hex = [...bytes].map(b => b.toString(16).padStart(2, '0')).join('');
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
}

export type AddWorkflowResult = { id: string } | { duplicateId: string };

export class WorkflowManager {
  private static STORAGE_KEY = 'trajectory-completed-history';
  private static MAX_COMPLETED = 50;
  private static MAX_TRACE_PER_WORKFLOW = 500;

  private _loaded: LoadedWorkflow[] = [];
  private _active: ActiveWorkflow[] = [];
  private _completed: CompletedWorkflow[] = [];
  private _focusedActiveId: string | null = null;
  private _snapshot: ManagerSnapshot = { loaded: [], active: [], completed: [], focusedActiveId: null };
  private _listeners = new Set<Listener>();
  private _coordinatorUnsubs = new Map<string, () => void>();
  /** Shared resource managers for environment-scoped cross-workflow sync, keyed by env OID set. */
  private _envResourceManagers = new Map<string, InMemoryResourceManager>();
  /** Guard against re-entrant pumping of sibling coordinators. */
  private _pumpingInProgress = false;

  constructor() {
    this._completed = WorkflowManager.loadCompleted();
    this._snapshot = { loaded: [], active: [], completed: this._completed, focusedActiveId: null };
  }

  private static loadCompleted(): CompletedWorkflow[] {
    try {
      const raw = localStorage.getItem(WorkflowManager.STORAGE_KEY);
      if (!raw) return [];
      return JSON.parse(raw) as CompletedWorkflow[];
    } catch {
      return [];
    }
  }

  private persistCompleted(): void {
    try {
      const capped = this._completed.slice(0, WorkflowManager.MAX_COMPLETED).map(wf => ({
        ...wf,
        trace: wf.trace.slice(0, WorkflowManager.MAX_TRACE_PER_WORKFLOW),
      }));
      localStorage.setItem(WorkflowManager.STORAGE_KEY, JSON.stringify(capped));
    } catch {
      // Storage full — silently fail
    }
  }

  getSnapshot = (): ManagerSnapshot => this._snapshot;

  subscribe = (listener: Listener): (() => void) => {
    this._listeners.add(listener);
    return () => this._listeners.delete(listener);
  };

  /** Load a workflow spec. Persists until explicitly removed. */
  addWorkflow(
    spec: MasterWorkflowSpecification,
    mediaMap: Record<string, string>,
    environments?: MasterEnvironmentLibrary[],
  ): AddWorkflowResult {
    const existing = this._loaded.find(w => w.specOid === spec.oid);
    if (existing) {
      return { duplicateId: existing.id };
    }

    const id = generateUUID();
    const loaded: LoadedWorkflow = {
      id,
      specOid: spec.oid,
      name: spec.description || spec.oid,
      localId: spec.local_id,
      version: spec.version,
      spec,
      mediaMap,
      environments: environments ?? [],
      loadedAt: Date.now(),
    };

    this._loaded = [...this._loaded, loaded];
    this.publish();
    return { id };
  }

  /** Remove a loaded workflow spec (manual delete only). */
  removeLoadedWorkflow(id: string): void {
    const loaded = this._loaded.find(w => w.id === id);
    if (loaded) {
      // Revoke blob URLs — no more instances can use them
      for (const url of Object.values(loaded.mediaMap)) {
        if (url.startsWith('blob:')) {
          URL.revokeObjectURL(url);
        }
      }
    }
    this._loaded = this._loaded.filter(w => w.id !== id);
    this.publish();
  }

  /** Get or create a shared resource manager for workflows sharing environments. */
  private getSharedResourceManager(environments: MasterEnvironmentLibrary[]): InMemoryResourceManager | undefined {
    if (environments.length === 0) return undefined;
    // Build a stable key from sorted environment OIDs
    const envOids = environments
      .flatMap(lib => lib.environment_specifications.map(e => e.oid))
      .sort()
      .join(',');
    if (!envOids) return undefined;
    let rm = this._envResourceManagers.get(envOids);
    if (!rm) {
      rm = new InMemoryResourceManager();
      this._envResourceManagers.set(envOids, rm);
    }
    return rm;
  }

  /**
   * Prepare a new workflow instance from a loaded spec: creates the coordinator,
   * calls coordinator.load(), adds to the active list, subscribes for completion —
   * but does NOT call coordinator.start(). Call runWorkflow(instanceId) to start.
   * Returns the new instanceId, or null if loadedId is not found.
   */
  prepareWorkflow(loadedId: string, startingParams?: Record<string, string>): string | null {
    const loaded = this._loaded.find(w => w.id === loadedId);
    if (!loaded) return null;

    const instanceId = generateUUID();
    const coordinator = new WorkflowCoordinator();

    // Deep-copy the spec so each instance has unique step OIDs
    const instanceSpec = deepCopySpec(loaded.spec);
    const setup = startingParams ? { starting_parameters: startingParams } : undefined;
    const sharedRM = this.getSharedResourceManager(loaded.environments);
    coordinator.load(instanceSpec, setup, { ...loaded.mediaMap }, loaded.environments, sharedRM);

    const active: ActiveWorkflow = {
      id: instanceId,
      sourceSpecId: loaded.id,
      name: loaded.name,
      localId: loaded.localId,
      version: loaded.version,
      coordinator,
      startedAt: Date.now(),
      instanceNumber: null,
    };

    this._active = [...this._active, active];
    this._focusedActiveId = instanceId;
    this.refreshInstanceNumbers();

    // Subscribe to detect completion and pump cross-workflow resource grants
    const unsub = coordinator.subscribe(() => {
      const snap = coordinator.getSnapshot();
      if (snap.workflowState === 'COMPLETED' || snap.workflowState === 'ABORTED'
        || snap.workflowState === 'STOPPED' || snap.workflowState === 'ERRORED') {
        queueMicrotask(() => {
          this.completeWorkflow(instanceId, snap);
        });
      }
      // Pump other active coordinators that share the same resource manager
      this.pumpSiblingCoordinators(instanceId);
    });
    this._coordinatorUnsubs.set(instanceId, unsub);

    this.publish();
    return instanceId;
  }

  /**
   * Start execution of a prepared workflow instance (call after prepareWorkflow and
   * any server binding setup). Pumps siblings after start.
   */
  runWorkflow(instanceId: string): void {
    const active = this._active.find(w => w.id === instanceId);
    if (!active) return;
    active.coordinator.start();
    this.pumpSiblingCoordinators(instanceId);
    this.publish();
  }

  /** Prepare and immediately start a workflow instance. Preserves original behavior. */
  startWorkflow(loadedId: string, startingParams?: Record<string, string>): string | null {
    const instanceId = this.prepareWorkflow(loadedId, startingParams);
    if (instanceId) this.runWorkflow(instanceId);
    return instanceId;
  }

  /** Move an active workflow to the completed list. */
  private completeWorkflow(instanceId: string, finalSnap: any): void {
    const active = this._active.find(w => w.id === instanceId);
    if (!active) return;

    // Unsubscribe from coordinator
    const unsub = this._coordinatorUnsubs.get(instanceId);
    if (unsub) {
      unsub();
      this._coordinatorUnsubs.delete(instanceId);
    }

    // Create completed entry with final state
    const completed: CompletedWorkflow = {
      id: active.id,
      sourceSpecId: active.sourceSpecId,
      name: active.name,
      localId: active.localId,
      version: active.version,
      finalState: finalSnap.workflowState,
      startedAt: active.startedAt,
      finishedAt: Date.now(),
      trace: finalSnap.trace ?? [],
      properties: finalSnap.properties ?? {},
      stepParams: finalSnap.stepParams ?? {},
    };

    // Remove from active
    this._active = this._active.filter(w => w.id !== instanceId);

    // Shift focus
    if (this._focusedActiveId === instanceId) {
      this._focusedActiveId = this._active.length > 0 ? this._active[0].id : null;
    }

    // Reset coordinator to free resources
    active.coordinator.reset();

    // Reassign instance numbers after removal
    this.refreshInstanceNumbers();

    // Add to completed (most recent first)
    this._completed = [completed, ...this._completed];
    this.publish();
    this.persistCompleted();
  }

  /** Remove a completed workflow from the history. */
  removeCompletedWorkflow(id: string): void {
    this._completed = this._completed.filter(w => w.id !== id);
    this.publish();
    this.persistCompleted();
  }

  /** Clear all completed workflows. */
  clearCompleted(): void {
    this._completed = [];
    this.publish();
    this.persistCompleted();
  }

  focusWorkflow(id: string): void {
    if (this._focusedActiveId === id) return;
    const exists = this._active.some(w => w.id === id);
    if (!exists) return;
    this._focusedActiveId = id;
    this.publish();
  }

  getCoordinator(id: string): WorkflowCoordinator | null {
    const active = this._active.find(w => w.id === id);
    return active?.coordinator ?? null;
  }

  /** Pump all active coordinators (except the source) to check for cross-workflow resource grants. */
  private pumpSiblingCoordinators(sourceId: string): void {
    if (this._pumpingInProgress) return;
    this._pumpingInProgress = true;
    try {
      for (const wf of this._active) {
        if (wf.id === sourceId) continue;
        wf.coordinator.pumpSharedResources();
      }
    } finally {
      this._pumpingInProgress = false;
    }
  }

  /** Release all resources across all active workflows. */
  releaseAllResources(): void {
    for (const wf of this._active) {
      wf.coordinator.releaseAllResources();
    }
  }

  /** Aggregate env-resource snapshot across all active workflows, deduplicated by
   *  composite resourceKey (`name` field). When the same env-scoped pool is shared
   *  by multiple workflows, the entry reflects current pool state once. */
  getEnvironmentResourceSnapshot(): import('@engine/types.js').ResourceSnapshotEntry[] {
    const seen = new Map<string, import('@engine/types.js').ResourceSnapshotEntry>();
    for (const wf of this._active) {
      for (const e of wf.coordinator.getEnvironmentResourceSnapshot()) {
        if (!seen.has(e.name)) seen.set(e.name, e);
      }
    }
    return [...seen.values()];
  }

  /** Aggregate workflow-scoped resource snapshot across all active workflows. */
  getWorkflowResourceSnapshot(): import('@engine/types.js').ResourceSnapshotEntry[] {
    const seen = new Map<string, import('@engine/types.js').ResourceSnapshotEntry>();
    for (const wf of this._active) {
      for (const e of wf.coordinator.getWorkflowResourceSnapshot()) {
        if (!seen.has(e.name)) seen.set(e.name, e);
      }
    }
    return [...seen.values()];
  }

  /** Reset all env-scoped resources across all active workflows. Each workflow
   *  triggers the same shared manager, so duplicates are harmless idempotent ops. */
  releaseAllEnvironmentResources(): string[] {
    const released = new Set<string>();
    for (const wf of this._active) {
      for (const r of wf.coordinator.releaseAllEnvironmentResources()) released.add(r);
    }
    return [...released];
  }

  /** Reset all workflow-scoped resources across all active workflows. */
  releaseAllWorkflowResources(): string[] {
    const released = new Set<string>();
    for (const wf of this._active) {
      for (const r of wf.coordinator.releaseAllWorkflowResources()) released.add(r);
    }
    return [...released];
  }

  /** Reset a single resource by composite resourceKey. Tries each coordinator until
   *  one reports the resource exists in its manager view. */
  resetResource(resourceKey: string): boolean {
    for (const wf of this._active) {
      if (wf.coordinator.resetResource(resourceKey)) return true;
    }
    return false;
  }

  /** Assign instance numbers when multiple instances of the same spec are active. */
  private refreshInstanceNumbers(): void {
    // Group active workflows by sourceSpecId
    const groups = new Map<string, ActiveWorkflow[]>();
    for (const wf of this._active) {
      const group = groups.get(wf.sourceSpecId) ?? [];
      group.push(wf);
      groups.set(wf.sourceSpecId, group);
    }
    for (const group of groups.values()) {
      if (group.length <= 1) {
        // Single instance — no number needed
        for (const wf of group) wf.instanceNumber = null;
      } else {
        // Multiple instances — assign sequential numbers by start order
        group.sort((a, b) => a.startedAt - b.startedAt);
        for (let i = 0; i < group.length; i++) {
          group[i].instanceNumber = i + 1;
        }
      }
    }
  }

  private publish(): void {
    this._snapshot = {
      loaded: this._loaded,
      active: this._active,
      completed: this._completed,
      focusedActiveId: this._focusedActiveId,
    };
    for (const listener of this._listeners) listener();
  }
}
